gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+<code>getcftypes</code>
+
+A full-node that receives a <code>getcftypes</code> message should respond with a
+<code>cftypes</code> message which is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| Var-Int
+| NumFilters
+| uint64
+| The number of supported filters.
+|-
+| NumFilters
+| SupportedFilters
+| [NumFilterBytes]byte
+| A byte slice with each byte denoting a supported filter type
+|}
+
+
+==== Compact Filter Header Chain ====
+
+As the filters described in this BIP ''are not'' consensus critical, meaning
+each filter is validated by full-nodes and committed into blocks by miners, we
+require an alternative (albeit less-binding) method to allow light clients to
+''identify'' and ''reject'' invalid filters. The purely p2p solution to this
+problem is to obtain a deterministic hash-chain of ''each'' filter. This hash
+chain or "filter header chain" is similar to the regular Bitcoin headers in
+that it allows a light client to verify the ''authenticity'' of a received
+filter.
+
+The filter header chain for a particular filter type is described by the
+following recurrence:
+<pre>
+filter_header(n: uint) -> [32]byte = 
+   let zero_hash [32]byte = {0*32}
+
+   if n == 0:
+       double-sha-256(genesis_block.prevblock || filter(0))
+
+   match filter(n):
+      case Some:
+          double-sha-256(filter_header(n-1) || double-sha-256(filter(n)))
+      case None:
+          double-sha-256(filter_header(n-1) || double-sha-256(zero_hash))
+
+   where filter(n) is the filter for block height n
+</pre>
+
+The filter header for the genesis block uses the hash stored in the prevblock
+field of the genesis block header itself, as there's no prior filter header
+(by definition).
+
+Due to the nature of filter construction, it's possible to construct a block
+such that an "empty" filter will be produced. This is the case of a coinbase
+transaction that has no data pushes in its public key script. In this case, the
+"hash" of said filter is simply "32 zeroes". 
+
+This filter header chain can be utilized by light clients to gain a greater
+degree of security against bamboozling full-nodes during their initial chain
+sync. In addition to fetching all the bitcoin headers, light clients
+implementing this BIP should also fetch all the ''filter headers'' from ''each'' of
+their connected peers. This allows light clients to efficiently detect nodes
+that advertise a conflicting filter chain history (by ensuring all nodes return
+the ''same'' filter header hash for a particular block hash).
+
+Additionally, the filter header chain also allows a light client to efficiently
+verify purported filter authenticity when fetching the next set of headers from
+chain tip. Instead of fetching the filter ''from each peer'' (which wastes
+bandwidth), a light client instead does the following:
+
+<pre>
+verify_from_tip(tip_block_hash: [32]byte):
```

Comment:
I dont know why this needs to be in the normative "Specification" section - clients can use this how they feel, and I think this is probably not the best approach anyway. You may wish to say something like clients SHOULD avoid syncing the full filter chain from all peers, but instead should only download from multiple peers when an inconsistency is detected.

Reply:
Moved to the "Implementation Notes" section.